src/backend/distributed/planner/distributed_planner.c

```diff
@@ -1112,23 +1112,40 @@ FinalizeDistributedPlan(DistributedPlan *plan, Query *originalQuery)
 static void
 RecordSubPlansUsedInPlan(DistributedPlan *plan, Query *originalQuery)
 {
-	/* first, get all the subplans in the query */
-	plan->usedSubPlanNodeList = FindSubPlansUsedInNode((Node *) originalQuery);
+	Node *havingQual = originalQuery->havingQual;
+
+	/* temporarily set to NULL, we're going to restore before the function returns */
+	originalQuery->havingQual = NULL;
```

Comment:
> Could you think of any other way of achieving the goal, other than mutating this object?

The simplest solution is to copy the originalQuery here again, which I want to avoid :) 

>  think this is the first time we are actually mutating the originalQuery which is the query postgres passes to the planned. This codepath is reached via recursion in distributed_planner with a PG_TRY/PG_CATCH construction.

As far as I can see, the source of `originalQuery` can be two different places:

- distributed_planner->CreateDistributedPlannedStmt->FinalizeDistributedPlan
- TryCreateDistributedPlannedStmt->CreateDistributedPlannedStmt->FinalizeDistributedPlan

In the first case, we're already using the copy of the parseTree: https://github.com/citusdata/citus/blob/master/src/backend/distributed/planner/distributed_planner.c#L211

In the second case, we copy it once more before try-catch: https://github.com/citusdata/citus/blob/master/src/backend/distributed/planner/distributed_planner.c#L764


Also, we always modify `originalQuery` in case of recursive planning, so should be fine doing it here as well. (https://github.com/citusdata/citus/blob/60a2bc5ec21bb53b95adc57a6d43cfe24759dbb8/src/backend/distributed/planner/recursive_planning.c#L189)

(Similarly there are parameters that we need to replace, we modify the original query: https://github.com/citusdata/citus/blob/4b5241c7b2a32bebc3ab5977d5c48291595f5e69/src/backend/distributed/planner/distributed_planner.c#L977)

What do you think? Looks OK to keep it as is

Reply:
given its copy and mutation in other places I think we can keep it like this here. The chances of it causing a problem are really slim to start with.